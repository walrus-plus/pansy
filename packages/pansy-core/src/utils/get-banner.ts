import { Banner } from '@pansy/types';
import { lodash } from '@walrus/shared-utils';
import stringifyAuthor from 'stringify-author';

/**
 * 获取Banner信息
 * @param banner banner配置
 * @param pkg package.json 中的数据
 */
export default (
  banner?: Banner,
  pkg?: { [k: string]: any }
): string => {
  if (!banner || lodash.isString(banner)) {
    return banner as string || '';
  }

  banner = { ...pkg, ...(banner === true ? {} : banner) };

  let author = '';

  if (lodash.isString(banner.author)) {
    author = banner.author;
  }

  if (lodash.isObject(banner.author)) {
    author = stringifyAuthor(banner.author)
  }

  const license = banner.license || '';

  return (
    '/*!\n' +
    ` * ${banner.name} v${banner.version}\n` +
    ` * (c) ${author || ''}\n` +
    (license && ` * Released under the ${license} License.\n`) +
    ' */'
  )
}
