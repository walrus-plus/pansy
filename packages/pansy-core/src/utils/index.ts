import { ModuleFormat as RollupFormat } from 'rollup'

export function getDefaultFileName(format: RollupFormat) {
  return format === 'cjs' ? `[name][min][ext]` : `[name].[format][min][ext]`
}
