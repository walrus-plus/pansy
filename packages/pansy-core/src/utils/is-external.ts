// @ts-ignore
import * as slash from 'slash';
import { lodash } from '@walrus/shared-utils';
import { Externals } from '@pansy/types';

export default function(externals: Externals, id: string, parentId?: string) {
  id = slash(id);

  if (!lodash.isArray(externals)) {
    externals = [externals] as Externals;
  }

  for (const external of externals) {
    if (
      lodash.isString(external) &&
      (id === external || id.includes(`/node_modules/${external}/`))
    ) {
      return true;
    }
    if (lodash.isRegExp(external)) {
      if (external.test(id)) {
        return true;
      }
    }
    if (typeof external === 'function') {
      if (external(id, parentId)) {
        return true;
      }
    }
  }

  return false;
}
