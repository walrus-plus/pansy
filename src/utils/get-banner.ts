import stringifyAuthor from 'stringify-author'

interface BannerInfo {
  /** Author name */
  name?: string
  /** package version */
  version?: string
  /** Author name or object */
  author?: any
  /** License name, like MIT */
  license?: string
}

export type Banner = string | BannerInfo | boolean

export default (banner?: Banner, pkg?: { [k: string]: any }): string => {
  if (!banner || typeof banner === 'string') {
    return banner || ''
  }

  banner = { ...pkg, ...(banner === true ? {} : banner) }

  const author =
    typeof banner.author === 'string'
      ? banner.author
      : typeof banner.author === 'object'
      ? stringifyAuthor(banner.author)
      : ''

  const license = banner.license || ''

  return (
    '/*!\n' +
    ` * ${banner.name} v${banner.version}\n` +
    ` * (c) ${author || ''}\n` +
    (license && ` * Released under the ${license} License.\n`) +
    ' */'
  )
}
