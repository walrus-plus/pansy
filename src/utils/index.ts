import { ModuleFormat } from 'rollup';
import { chalk as colors } from '@walrus/shared-utils';
import prettyBytes from 'pretty-bytes';
import boxen from 'boxen';
import textTable from 'text-table';
import stringWidth from 'string-width';
import logger from '../logger';

export interface Asset {
  absolute: string;
  source: string;
}

export type Assets = Map<string, Asset>;

export function getDefaultFileName(format: ModuleFormat) {
  return format === 'cjs' ? `[name][min][ext]` : `[name].[format][min][ext]`;
}

export async function printAssets(assets: Assets, title: string) {
  const gzipSize = await import('gzip-size').then((res) => res.default);
  const table = await Promise.all(
    [...assets.keys()].map(async (relative) => {
      const asset = assets.get(relative) as Asset;
      const size = asset.source.length;
      return [colors.green(relative), prettyBytes(size), prettyBytes(await gzipSize(asset.source))];
    })
  );
  table.unshift(['File', 'Size', 'Gzipped'].map((v) => colors.dim(v)));
  logger.success(title);
  logger.log(
    boxen(
      textTable(table, {
        stringLength: stringWidth
      })
    )
  );
}
