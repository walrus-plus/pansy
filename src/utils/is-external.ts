import slash from 'slash'
import { NormalizedConfig } from '../'

export default function(
  externals: NormalizedConfig['externals'],
  id: string,
  parentId?: string
) {
  id = slash(id)

  if (!Array.isArray(externals)) {
    externals = [externals] as NormalizedConfig['externals']
  }

  for (const external of externals) {
    if (
      typeof external === 'string' &&
      (id === external || id.includes(`/node_modules/${external}/`))
    ) {
      return true
    }
    if (external instanceof RegExp) {
      if (external.test(id)) {
        return true
      }
    }
    if (typeof external === 'function') {
      if (external(id, parentId)) {
        return true
      }
    }
  }

  return false
}
