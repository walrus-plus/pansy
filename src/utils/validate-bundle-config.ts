import signale from 'signale';
import { join } from 'path';
import isString from '@pansy/is-string';
import isArray from '@pansy/is-array';
import isObject from '@pansy/is-object';
import { existsSync } from 'fs';
import { chalk } from '@walrus/shared-utils';
import { isTypescriptFile } from './';
import { NormalizedConfig } from '../types';

export default function(config: NormalizedConfig, rootDir: string) {
  const { output, input } = config;

  if (output && output.format) {
    // 检验 cjs、commonjs是否重复设置
    if (output.format.indexOf('cjs') !== -1 && output.format.indexOf('commonjs') !== -1) {
      signale.warn(`${chalk.cyan('cjs')} and ${chalk.cyan('commonjs')} only need to set one`);
    }
    // 检验 es、esm是否重复设置
    if (output.format.indexOf('es') !== -1 && output.format.indexOf('esm') !== -1) {
      signale.warn(`${chalk.cyan('es')} and ${chalk.cyan('esm')} only need to set one`);
    }

    // umd格式必须制定output.name
    if (output.format.indexOf('umd') !== -1) {
      throw new Error(`${chalk.cyan('umd')} format must be set ${chalk.cyan('output.name')}`);
    }
  }

  const tsConfigPath = join(rootDir, 'tsconfig.json');
  const tsConfig = existsSync(tsConfigPath);

  let inputs: string[] = [];

  //  解析入口文件为数组
  if (input) {
    if (isString(input)) {
      inputs = [input];
    }
    if (isObject(input)) {
      Object.keys(input).forEach(key => {
        // @ts-ignore
        inputs.push(input[key]);
      })
    }
    if (isArray(config.input)) {
      inputs = config.input as string[];
    }
  }

  // ts项目没有配置tsconfig.json
  if (!tsConfig && inputs.some(isTypescriptFile as any)) {
    signale.info(
      `Project using ${chalk.cyan('typescript')} but tsconfig.json not exists. Use default config.`
    );
  }
}
