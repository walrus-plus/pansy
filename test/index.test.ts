import { snapshot, fixture } from './utils';

snapshot({
  title: 'defaults',
  input: 'index.js',
  cwd: fixture('defaults')
});

snapshot(
  {
    title: 'exclude file',
    input: 'index.js',
    cwd: fixture('exclude-file')
  },
  {
    externals: ['./foo.js']
  }
);

snapshot(
  {
    title: 'extendOptions',
    input: ['foo.js', 'bar.js'],
    cwd: fixture('extend-options')
  },
  {
    output: {
      format: ['umd', 'umd-min', 'cjs']
    },
    extendConfig(config, { format }) {
      if (format === 'umd') {
        config.output.moduleName = 'umd';
      }
      if (format.endsWith('-min')) {
        config.output.moduleName = 'min';
      }
      return config;
    }
  }
);

snapshot(
  {
    title: 'bundle-node-modules',
    input: 'index.js',
    cwd: fixture('bundle-node-modules')
  },
  {
    bundleNodeModules: true
  }
);

snapshot({
  title: 'async',
  input: 'index.js',
  cwd: fixture('async')
});

snapshot(
  {
    title: 'uglify',
    input: 'index.js',
    cwd: fixture('uglify')
  },
  {
    output: {
      format: 'cjs-min'
    }
  }
);

snapshot(
  {
    title: 'inline-certain-modules',
    input: 'index.js',
    cwd: fixture('inline-certain-modules')
  },
  {
    bundleNodeModules: ['fake-module']
  }
);

snapshot(
  {
    title: 'target:browser',
    input: 'index.js',
    cwd: fixture('target/browser')
  },
  {
    output: {
      target: 'browser'
    }
  }
);
