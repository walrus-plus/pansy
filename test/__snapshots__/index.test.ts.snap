// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`async: async dist/index.js 1`] = `
"'use strict';

function _async(f) {
  return function () {
    for (var args = [], i = 0; i < arguments.length; i++) {
      args[i] = arguments[i];
    }

    try {
      return Promise.resolve(f.apply(this, args));
    } catch (e) {
      return Promise.reject(e);
    }
  };
}

const a = {
  a: 1
};
var index = _async(function () {
  return Object.assign({}, a, {
    b
  });
});

module.exports = index;
"
`;

exports[`bundle-node-modules: bundle-node-modules dist/index.js 1`] = `
"'use strict';

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var all = _interopDefault(require('some-package-that-does-not-exists'));

var pokemon = ['test'];

var index = {
  all,
  pokemon
};

module.exports = index;
"
`;

exports[`custom rollup plugin: custom rollup plugin dist/index.js 1`] = `
"'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const tt = 'is there anything?';

function afun() {
  return tt;
}

function another() {
  return afun();
}

exports.another = another;
"
`;

exports[`defaults: defaults dist/index.js 1`] = `
"'use strict';

var index = 42;

module.exports = index;
"
`;

exports[`exclude file: exclude file dist/index.js 1`] = `
"'use strict';

var foo = 'foo';

module.exports = foo;
"
`;

exports[`extendOptions: extendOptions dist/bar.js 1`] = `
"'use strict';

var bar = 'bar';

module.exports = bar;
"
`;

exports[`extendOptions: extendOptions dist/bar.umd.js 1`] = `
"(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory() :
	typeof define === 'function' && define.amd ? define(factory) :
	(global = global || self, global.umd = factory());
}(this, (function () { 'use strict';

	var bar = 'bar';

	return bar;

})));
"
`;

exports[`extendOptions: extendOptions dist/bar.umd.min.js 1`] = `
"!function(e,n){\\"object\\"==typeof exports&&\\"undefined\\"!=typeof module?module.exports=n():\\"function\\"==typeof define&&define.amd?define(n):(e=e||self).min=n()}(this,(function(){\\"use strict\\";return\\"bar\\"}));
"
`;

exports[`extendOptions: extendOptions dist/foo.js 1`] = `
"'use strict';

var foo = 'foo';

module.exports = foo;
"
`;

exports[`extendOptions: extendOptions dist/foo.umd.js 1`] = `
"(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory() :
	typeof define === 'function' && define.amd ? define(factory) :
	(global = global || self, global.umd = factory());
}(this, (function () { 'use strict';

	var foo = 'foo';

	return foo;

})));
"
`;

exports[`extendOptions: extendOptions dist/foo.umd.min.js 1`] = `
"!function(e,n){\\"object\\"==typeof exports&&\\"undefined\\"!=typeof module?module.exports=n():\\"function\\"==typeof define&&define.amd?define(n):(e=e||self).min=n()}(this,(function(){\\"use strict\\";return\\"foo\\"}));
"
`;

exports[`inline-certain-modules: inline-certain-modules dist/index.js 1`] = `
"'use strict';

require('another-module');

var fakeModule = a => a;

var index = fakeModule(1);

module.exports = index;
"
`;

exports[`target:browser: target:browser dist/index.js 1`] = `
"'use strict';

var foo = 'browser';

module.exports = foo;
"
`;

exports[`uglify: uglify dist/index.min.js 1`] = `
"\\"use strict\\";Object.defineProperty(exports,\\"__esModule\\",{value:!0});const a=1;exports.a=1;
"
`;
